File: C/sorts.c
#include <stdio.h>

void bubbleSort (int nums[], int size) {
  int swap = 0;
  for (int i=0; i<size; i++) {
    for (int j=0; j<size-i-1; j++) {
      if (nums[j] > nums[j+1]) {
        int tmp = nums[j];
        nums[j] = nums[j+1];
        nums[j+1] = tmp;
        swap = 1;
      }
    }
    if (!swap) {
      break;
    }
  }
}

void selectionSort (int nums[], int size) {
  for (int i=0; i<size; i++) {
    int min_idx = i;
    for (int j=i+1; j<size; j++) {
      if (nums[j] < nums[min_idx]) {
        min_idx = j;
      }
    }
    int tmp = nums[i];
    nums[i] = nums[min_idx];
    nums[min_idx] = tmp;
  }
}

void insertionSort (int nums[], int size) {
  for (int i=1; i<size; i++) {
    int tmp = nums[i];
    int j = i-1;
    while (j>=0 && nums[j] > tmp) {
      nums[j+1] = nums[j];
      j--; 
    }
    nums[j+1] = tmp;
  }
}

void merge (int nums[], int low, int mid, int high) {
    
    int temp[high+1];
    int left = low;
    int right = mid+1;
    int i = 0;
    // [low..mid]
    // [mid+1..high]
    while (left <= mid && right <= high) {
        if (nums[left] <= nums[right]) {
            temp[i++] = nums[left++];
        } else {
            temp[i++] = nums[right++];
        }
    }

    while (left <= mid) {
        temp[i++] = nums[left++];
    }
    
    while (right <= high) {
        temp[i++] = nums[right++];
    }

    // copy temp to original array
    for (int i = low; i <= high; i++) {
        nums[i] = temp[i-low];
    }
}

void mergeSort (int nums[], int low, int high) {

    // base case
    if (low >= high) {
        return;
    }
    int mid = low + (high - low) / 2;
    mergeSort(nums, low, mid);
    mergeSort(nums, mid+1, high);
    merge(nums, low, mid, high);
}

int partition (int nums[], int low, int high) {

    int pivot = nums[low]; // chosen pivot
    int i = low;
    int j = high;
    while (i < j) {
        while (nums[i] <= pivot && i <= high - 1) {
            i++;
        }
        while (nums[j] > pivot && j >= low + 1) {
            j--;
        }
        if (i < j) {
            int temp = nums[i];
            nums[i] = nums[j];
            nums[j] = temp;
        }
    }

    // swap pivot to correct place
    int temp = nums[low];
    nums[low] = nums[j];
    nums[j] = temp;

    return j;
}

void quickSort (int nums[], int low, int high) {

    if (low < high) {
        int p_index = partition(nums, low, high);
        quickSort(nums, low, p_index - 1);
        quickSort(nums, p_index + 1, high);
    }
}

void print_array(int array[], int size) {
  for (int i=0; i<size; i++) {
    printf("%d ", array[i]);
  }
  printf("\n");
}

int main () {

  int nums[10] = {4,6,5,1,3,2,7,9,10,8};
  int size = sizeof(nums)/sizeof(nums[0]);

  print_array(nums, size);
  //bubbleSort(nums, size);
  //selectionSort(nums, size);
  //insertionSort(nums, size);
  int low = 0; int high = size-1;
  mergeSort(nums, low, high);
  //quickSort(nums, low, high);
  print_array(nums, size);

  return 0;
}

File: python/sorts.py
# collection of a few different sorting algorithms

class Sorts:

    # shrink inner for loop by len of array -i -1 to not do many pointless iterations once the sorting is finished (can also use boolean as show here - most basic swap/algo
    def bubbleSort(self, nums):
        for i in range(len(nums)):
            swap = False
            for j in range(len(nums)-i-1):
                if nums[j] > nums[j+1]:
                    nums[j], nums[j+1] = nums[j+1], nums[j]
                    swap = True
            if not swap:
                break
        return nums

    # start inner loop index+1 to len of array and keep track of minimum index throughout each iteration - perform swap after each outter loop iteration
    def selectionSort(self, nums):
        for i in range(len(nums)):
            min_idx = i
            for j in range(i+1, len(nums)):
                if nums[j] < nums[min_idx]:
                    min_idx = j
            nums[i], nums[min_idx] = nums[min_idx], nums[i]
        return nums

    # start at index=1 and compare everything to the left in the array - then swap if condition
    def insertionSort(self, nums):
        for i in range(1, len(nums)):
            temp = nums[i]
            j = i - 1
            while j >= 0 and nums[j] > temp:
                nums[j+1] = nums[j]
                j -= 1
            nums[j+1] = temp
        return nums

    # divide and merge algorithm -
    def mergeSort(self, nums, low, high):

        def merge(nums, low, mid, high):
            temp = [0]*(high+1)
            l, r = low, mid+1
            i = 0
            while l <= mid and r <= high:
                if nums[l] <= nums[r]:
                    temp[i] = nums[l]
                    l += 1
                    i += 1
                else:
                    temp[i] = nums[r]
                    r += 1
                    i += 1

            while l <= mid:
                temp[i] = nums[l]
                l += 1
                i += 1
            while r <= high:
                temp[i] = nums[r]
                r += 1
                i += 1

            for i in range(low, high+1):
                nums[i] = temp[i-low]

        if low >= high: return # base case

        mid = low + (high - low) // 2
        self.mergeSort(nums, low, mid)
        self.mergeSort(nums, mid+1, high)
        merge(nums, low, mid, high)
            
        return nums


    # divide and conquer algorithm -
    def quickSort(self, nums, low, high):

        def partition(nums, low, high):
            pivot = nums[low]
            i = low
            j = high
            while i < j:
                while nums[i] <= pivot and i <= high - 1:
                    i += 1
                while nums[j] > pivot and j >= low + 1:
                    j -= 1
                if i < j:
                    nums[i], nums[j] = nums[j], nums[i]

            nums[low], nums[j] = nums[j], nums[low]

            return j
        if low < high:
            p_index = partition(nums, low, high)
            self.quickSort(nums, low, p_index-1)
            self.quickSort(nums, p_index+1, high)
        
        return nums


import random
if __name__ == "__main__":

    nums = [x+1 for x in range(10)]
    random.shuffle(nums)

    print(nums)
    #print(Sorts().bubbleSort(nums))
    #print(Sorts().selectionSort(nums))
    #print(Sorts().insertionSort(nums))
    low, high = 0, len(nums)-1
    print(Sorts().mergeSort(nums, low, high))
    #print(Sorts().quickSort(nums, low, high))
